### PR TITLE
[FEATURE] `UnexpectedRowsExpectation`

### PR DESCRIPTION
**Pull request by cdkini**
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: tests/expectations/test_expectation.py [hunk 1]
```diff
@@ -251,3 +257,58 @@ def test_unrecognized_expectation_arg_raises_error():
...
+    )
+    result = batch.validate(expectation)
+
+    # DB contains a single deviation (passenger_count == 7)
+    assert result.success is False
+
+    res = result.result
+    assert res["observed_value"] == 1
+
+    details = res["details"]
```
Comment (tyler-hoffman) [inline tests/expectations/test_expectation.py]: nit: Spreading this out into 3 lines with variables feels like it makes this harder to read that just `res["details"]["value_counts"]["unexpected_rows"]`. Also asserting on what that unexpected row is would help folks reading the test to understand what the output actually is.

File: tests/expectations/test_expectation.py [hunk 2]
```diff
@@ -251,3 +257,58 @@ def test_unrecognized_expectation_arg_raises_error():
...
+    in_memory_runtime_context: AbstractDataContext, taxi_db_path: str
+) -> SqliteDatasource:
+    context = in_memory_runtime_context
+    datasource_name = "my_sqlite_datasource"
+    return context.sources.add_sqlite(
+        datasource_name, connection_string=f"sqlite:///{taxi_db_path}"
+    )
+
+
+@pytest.mark.sqlite
```
Comment (tyler-hoffman) [inline tests/expectations/test_expectation.py]: I'd like to see a test on the success case as well

File: great_expectations/expectations/expectation.py [hunk 3]
```diff
@@ -1592,6 +1592,51 @@ def _validate_metric_value_between(  # noqa: PLR0912
...
+    def _validate_query(cls, query: str) -> str:
+        is_valid = "active_batch" in query
+        try:
+            # Using a dummy value to check syntactic correctness
+            substituted_query = query.format(active_batch="dummy_value")
+            # If the query is not parameterized, it will be the same as the substituted query
+            if query == substituted_query:
+                is_valid = False
+        except KeyError:
+            is_valid = False
```
Comment (billdirks) [inline great_expectations/expectations/expectation.py]: In the import section do `from string import Formatter` then here do:
```
is_valid = "active_batch" in [f[1] for f in Formatter().parse(query)]
```
See docs [here](https://docs.python.org/3/library/string.html#string.Formatter.parse). Python needs to be able to parse format strings and provides the same capability to developers. You can leverage this so you don't have to do these checks here.

To see what parse is doing you can iterate and print the elements of `Formatter().parse(query)` to see what it returns. The first element is a string with the text before a format variable `{name}` and the second element (that I grab in the snippet above) is the `name` of the format variable. The docs say a bit more.

File: tests/expectations/test_expectation.py [hunk 4]
```diff
@@ -251,3 +267,107 @@ def test_unrecognized_expectation_arg_raises_error():
...
+        pytest.param(
+            "SELECT * FROM {active_batch} WHERE passenger_count > 7",
+            True,
+            0,
+            [],
+            id="success",
+        ),
+        pytest.param(
+            # There is a single instance where passenger_count == 7
+            "SELECT * FROM {active_batch} WHERE passenger_count > 6",
```
Comment (jcampbell) [inline tests/expectations/test_expectation.py:324]: Didn't you mention we're using {batch} instead of {active_batch}?

Comment (cdkini) [reply tests/expectations/test_expectation.py:324]: yup that impacts more than this expectation so I'll have to follow up with another PR to change things across the board

File: great_expectations/expectations/metrics/query_metrics/query_table.py [hunk 5]
```diff
@@ -22,7 +22,25 @@
 
 class QueryTable(QueryMetricProvider):
     metric_name = "query.table"
-    value_keys = ("query",)
+    value_keys = (
+        "query",
```
Comment (jcampbell) [inline great_expectations/expectations/metrics/query_metrics/query_table.py:26]: Any reason not to remove query here?

File: great_expectations/expectations/expectation.py [hunk 6]
```diff
@@ -1592,6 +1593,53 @@ def _validate_metric_value_between(  # noqa: PLR0912
...
+        metrics: dict,
+        runtime_configuration: dict | None = None,
+        execution_engine: ExecutionEngine | None = None,
+    ) -> Union[ExpectationValidationResult, dict]:
+        metric_value = metrics["query.table"]
+        return {
+            "success": len(metric_value) == 0,
+            "result": {
+                "observed_value": len(metric_value),
+                "details": {"value_counts": {"unexpected_rows": metric_value}},
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py]: Can we move this out of `value_counts`?

**Review by tyler-hoffman — APPROVED** [review]

File: great_expectations/expectations/metrics/query_metrics/query_table.py (hunk 5 above)
Comment (cdkini) [reply great_expectations/expectations/metrics/query_metrics/query_table.py:26]: we currently utilize it in other expectations. I can update things across the board but that'll increase scope so I'll opt for a separate PR

**Pull request merged**